### PR TITLE
[SC-838] Add ERC4626 support to ForeignController with Morpho Testing

### PR DESCRIPTION
**Pull request by hexonaut**
Add ERC4626 support to ForeignController along with Morpho Testing.

### CODE REVIEW

File: src/ForeignController.sol [hunk 1]
```diff
@@ -214,6 +216,67 @@ contract ForeignController is AccessControl {
...
+    /**********************************************************************************************/
+    /*** Relayer ERC4626 functions                                                              ***/
+    /**********************************************************************************************/
+
+    function depositERC4626(address vault, uint256 amount)
+        external
+        onlyRole(RELAYER)
+        isActive
+        rateLimited(
+            RateLimitHelpers.makeAssetKey(LIMIT_VAULT_DEPOSIT, vault),
```
Comment (lucas-manuel) [inline src/ForeignController.sol:228]: We should use another term instead of vault across the code here to avoid conflating with the allocation system, I used `token` in the mainnet implementation, not sure what you think of that

File: test/base-fork/Morpho.t.sol [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+    function test_morpho_deposit_rateLimited() external {
+        deal(Base.USDS, address(almProxy), 25_000_001e18);
+
+        assertEq(usdsVault.convertToAssets(usdsVault.balanceOf(address(almProxy))), 0);
+        assertEq(IERC20(Base.USDS).balanceOf(address(almProxy)),                    25_000_001e18);
+
+        vm.expectRevert("RateLimits/rate-limit-exceeded");
+        vm.prank(relayer);
+        foreignController.depositERC4626(MORPHO_VAULT_USDS, 25_000_001e18);
```
Comment (lucas-manuel) [inline test/base-fork/Morpho.t.sol:127]: `25_000_000e18 + 1`, applies to all boundary conditions

File: test/base-fork/Morpho.t.sol [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    function test_morpho_deposit_frozen() external {
+        vm.prank(freezer);
+        foreignController.freeze();
+
+        vm.prank(relayer);
+        vm.expectRevert("ForeignController/not-active");
+        foreignController.depositERC4626(MORPHO_VAULT_USDS, 1_000_000e18);
+    }
+
+    function test_morpho_deposit_rateLimited() external {
```
Comment (lucas-manuel) [inline test/base-fork/Morpho.t.sol:119]: test_morpho_deposit_rateLimitedBoundary, should add one for USDC as well here

File: test/base-fork/Morpho.t.sol [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+        deal(Base.USDS, address(almProxy), 25_000_001e18);
+
+        assertEq(usdsVault.convertToAssets(usdsVault.balanceOf(address(almProxy))), 0);
+        assertEq(IERC20(Base.USDS).balanceOf(address(almProxy)),                    25_000_001e18);
```
Comment (lucas-manuel) [inline test/base-fork/Morpho.t.sol:123]: Don't need state assertions in boundary failure mode tests

File: test/base-fork/Morpho.t.sol [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+        vm.expectRevert("RateLimits/rate-limit-exceeded");
+        vm.prank(relayer);
+        foreignController.depositERC4626(MORPHO_VAULT_USDS, 25_000_001e18);
+
+        vm.prank(relayer);
+        foreignController.depositERC4626(MORPHO_VAULT_USDS, 25_000_000e18);
+
+        assertEq(usdsVault.convertToAssets(usdsVault.balanceOf(address(almProxy))), 25_000_000e18);
+        assertEq(IERC20(Base.USDS).balanceOf(address(almProxy)),                    1e18);
+    }
```
Comment (lucas-manuel) [inline test/base-fork/Morpho.t.sol:134]: Here's how the test should look:

```solidity
    function test_morpho_deposit_rateLimitedBoundary() external {
        deal(Base.USDS, address(almProxy), 25_000_000e18 + 1);

        vm.expectRevert("RateLimits/rate-limit-exceeded");
        vm.startPrank(relayer);
        foreignController.depositERC4626(MORPHO_VAULT_USDS, 25_000_000e18 + 1);

        foreignController.depositERC4626(MORPHO_VAULT_USDS, 25_000_000e18);
    }
```

File: test/base-fork/Morpho.t.sol [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+        return 22841965;  // November 24, 2024
+    }
+
+}
+
+/**********************************************************************************************/
+/*** Only testing USDS failure modes because it is the same for USDC                        ***/
+/**********************************************************************************************/
+
+contract MorphoFailureTests is MorphoBaseTest {
```
Comment (lucas-manuel) [inline test/base-fork/Morpho.t.sol:99]: Break out contracts by function, deposit, withdraw, and redeem failures

File: test/base-fork/Morpho.t.sol [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+        foreignController.redeemERC4626(MORPHO_VAULT_USDS, 1_000_000e18);
+    }
+
+}
+
+/**********************************************************************************************/
+/*** Success modes testing both USDS and USDC                                               ***/
+/**********************************************************************************************/
+
+contract MorphoUSDSSuccessTests is MorphoBaseTest {
```
Comment (lucas-manuel) [inline test/base-fork/Morpho.t.sol:178]: Break out by function, doing usds and usdc side by side

Comment (lucas-manuel) [reply test/base-fork/Morpho.t.sol:178]: MorphoDepositSuccessTests etc

File: test/base-fork/Morpho.t.sol [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+
+    function test_morpho_usds_deposit() public {
+        deal(Base.USDS, address(almProxy), 1_000_000e18);
+
+        assertEq(usdsVault.convertToAssets(usdsVault.balanceOf(address(almProxy))), 0);
```
Comment (lucas-manuel) [inline test/base-fork/Morpho.t.sol:183]: Add assertions for approvals being zero before and after

File: test/base-fork/Morpho.t.sol [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+        assertEq(IERC20(Base.USDS).balanceOf(address(almProxy)),                    1_000_000e18);
+
+        vm.prank(relayer);
+        foreignController.depositERC4626(MORPHO_VAULT_USDS, 1_000_000e18);
```
Comment (lucas-manuel) [inline test/base-fork/Morpho.t.sol:187]: Assert return values for all function calls

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: test/base-fork/Morpho.t.sol (hunk 2 above)
Comment (hexonaut) [reply test/base-fork/Morpho.t.sol:127]: updated

File: test/base-fork/Morpho.t.sol (hunk 3 above)
Comment (hexonaut) [reply test/base-fork/Morpho.t.sol:119]: added

File: test/base-fork/Morpho.t.sol (hunk 4 above)
Comment (hexonaut) [reply test/base-fork/Morpho.t.sol:123]: removed

File: test/base-fork/Morpho.t.sol (hunk 5 above)
Comment (hexonaut) [reply test/base-fork/Morpho.t.sol:134]: changed

File: test/base-fork/Morpho.t.sol (hunk 6 above)
Comment (hexonaut) [reply test/base-fork/Morpho.t.sol:99]: at a certain point this seems excessive having 1-2 tests per contract

File: test/base-fork/Morpho.t.sol (hunk 7 above)
Comment (hexonaut) [reply test/base-fork/Morpho.t.sol:178]: same here, i think there is a point of diminishing returns having 1 test per contract

File: test/base-fork/Morpho.t.sol (hunk 8 above)
Comment (hexonaut) [reply test/base-fork/Morpho.t.sol:183]: done

File: test/base-fork/Morpho.t.sol (hunk 9 above)
Comment (hexonaut) [reply test/base-fork/Morpho.t.sol:187]: added

**Review by lucas-manuel — APPROVED** [review]
